target/scripts/helpers/error.sh

```diff
@@ -49,34 +50,49 @@ function dms_panic
       SHUTDOWN_MESSAGE="Invalid value for ${PANIC_INFO}!"
       ;;
 
+    ( 'general' )
+      SHUTDOWN_MESSAGE=${PANIC_INFO}
+      ;;
+
     ( * ) # `dms_panic` was called directly without a valid PANIC_TYPE
       SHUTDOWN_MESSAGE='Something broke :('
       ;;
   esac
 
   if [[ -n ${PANIC_SCOPE:-} ]]
   then
-    _shutdown "${PANIC_SCOPE} | ${SHUTDOWN_MESSAGE}"
+    _shutdown "${PANIC_SCOPE} | ${SHUTDOWN_MESSAGE}" "${PANIC_STRATEGY}"
   else
-    _shutdown "${SHUTDOWN_MESSAGE}"
+    _shutdown "${SHUTDOWN_MESSAGE}" "${PANIC_STRATEGY}"
   fi
 }
 
 # Convenience wrappers based on type:
-function dms_panic__fail_init { dms_panic 'fail-init' "${1}" "${2}"; }
-function dms_panic__no_env { dms_panic 'no-env' "${1}" "${2}"; }
-function dms_panic__no_file { dms_panic 'no-file' "${1}" "${2}"; }
-function dms_panic__misconfigured { dms_panic 'misconfigured' "${1}" "${2}"; }
-function dms_panic__invalid_value { dms_panic 'invalid-value' "${1}" "${2}"; }
+function _dms_panic__fail_init     { dms_panic 'fail-init'     "${1:-}" "${2:-}" "${3:-}" ; }
+function _dms_panic__no_env        { dms_panic 'no-env'        "${1:-}" "${2:-}" "${3:-}" ; }
+function _dms_panic__no_file       { dms_panic 'no-file'       "${1:-}" "${2:-}" "${3:-}" ; }
+function _dms_panic__misconfigured { dms_panic 'misconfigured' "${1:-}" "${2:-}" "${3:-}" ; }
+function _dms_panic__invalid_value { dms_panic 'invalid-value' "${1:-}" "${2:-}" "${3:-}" ; }
+function _dms_panic__general       { dms_panic 'general'       "${1:-}" "${2:-}" "${3:-}" ; }
 
 # Call this method when you want to panic (i.e. emit an 'ERROR' log, and exit uncleanly).
 # `dms_panic` methods should be preferred if your failure type is supported.
 function _shutdown
 {
-  _log 'error' "${1}"
+  _log 'error' "${1:-_shutdown called without message}"
   _log 'error' 'Shutting down'
 
   sleep 1
   kill 1
-  exit 1
+
+  if [[ ${2:-wait} == 'immediate' ]]
+  then
+    # In case the user requested an immediate exit, he ensure he is not in a subshell
+    # call and exiting the whole script is safe. This way, we make the shutdown quicker.
+    exit 1
+  else
+    # We can simply wait until Supervisord has terminated all processes; this way,
+    # we do not return from a subshell call and continue as if nothing happened.
+    sleep 1000
+  fi
```

Comment:
Would it not be cleaner to approach with the first panic call (`_shutdown()` rather) creating some state like `/tmp/panic`, and just checking if this file exists to block with the `sleep 1000`?

No need for the `immediate` arg then, which avoids calls with empty arg before it for scope.

At least that seems simpler and pretty much always the behaviour desired? (_`exit 1` first call, any subsequent calls block_)

In the [cited failure case](https://github.com/docker-mailserver/docker-mailserver/issues/3133), the log should stop where it expected from the looks of it.

I understand the drawback would be scripts continue to run after the first `exit 1`? But I don't quite understand how that `exit 1` interacts with the scope of the `ssl.sh` helper, these are all panic calls within the same method, so what happens if `sleep 1000` was at the end of the panic call? It always hits the `sleep 1000`?

We are avoiding the `immediate` arg only for `ssl.sh` and `dns.sh`, if keeping the arg approach, it may be simpler to invert this arg to opt-in to blocking instead?

Reply:
I chose not to take the file approach because I really dislike scripts continuing after `_shutdown` was called (potentially causing behaviour we do not expect at all). I would definitely avoid going to UB-lane.

This is also the reason for the arg dance; using `sleep` is the conservative but also very safe method. Hence, it should be the default. I am not a big fan of `''` args either, but it's working fine.

I'd keep it as is for now. If we ever feel the need to tackle this again, I will assist.